datastore/longhorn.go

```diff
@@ -490,11 +490,26 @@ func (s *DataStore) ValidateV2DataEngineEnabled(dataEngineEnabled bool) (ims []*
 }
 
 func (s *DataStore) AreAllRWXVolumesDetached() (bool, error) {
-	pods, err := s.ListShareManagerPodsRO("")
+	volumes, err := s.ListVolumesRO()
 	if err != nil {
 		return false, err
 	}
-	return len(pods) == 0, nil
+	for _, volume := range volumes {
+		if volume.Spec.AccessMode != longhorn.AccessModeReadWriteMany {
+			continue
+		}
+
+		volumeAttachment, err := s.GetLHVolumeAttachmentByVolumeName(volume.Name)
```

Comment:
As far as I can tell, this change will not help with https://github.com/longhorn/longhorn/issues/9022. The issue @PhanLe1010 and @james-munson are having (if I understand it properly) is that NFS clients still exist on the cluster trying to use the old ClusterIP. The `volumeattachment.longhorn.io` does not know anything about them. It is only concerned with the underlying Longhorn volume. In order to ensure all NFS clients are gone, I think we must look at `volumeattachment.storage.k8s.io` objects. Alternatively, do we trust the `volume.status.KubernetesStatus`? Reading through the code, I think it is the case that all RWX clients are gone if `lastPodRefAt != ""`. It indicates that we can no longer find any pods on the cluster that reference the volume. So either they have been successfully torn down or force deleted (in which case, all bets are off).

https://github.com/longhorn/longhorn-manager/blob/b615732bc75d5513a7f3289a667af1f06a55af1d/controller/kubernetes_pv_controller.go#L460-L467

Reply:
Sorry, we can disregard this. I was confused, but @PhanLe1010 set me straight offline.

We actually DO create a `volumeattachment.longhorn.io` for every node attempting to use an RWX volume. In ADDITION, we have the one responsible for attaching the underlying volume to the share manager's node.

https://github.com/longhorn/longhorn-manager/blob/b615732bc75d5513a7f3289a667af1f06a55af1d/controller/volume_attachment_controller.go#L579-L583

If there are no attachment tickets, it means Kubernetes has issued ControllerUnpublishVolume RPCs for every node that had the volume RWX volume attached. It should only do this if NodeUnpublishVolume and NodeUnstageVolume RPCs have succeeded. These RPCs should clean up all active NFS clients.